utils/cmd/cmd.c

```diff
@@ -0,0 +1,721 @@
+/*
+ * BlueALSA - cmd.c
+ * Copyright (c) 2016-2021 Arkadiusz Bokowy
+ *
+ * This file is a part of bluez-alsa.
+ *
+ * This project is licensed under the terms of the MIT license.
+ *
+ */
+
+
+#if HAVE_CONFIG_H
+# include <config.h>
+#endif
+
+#include "shared/dbus-client.h"
+#include "shared/defs.h"
+#include "../dbus.h"
+
+#include <errno.h>
+#include <getopt.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+
+static int list_pcms(int argc, char *argv[]);
+static int get_codecs(int argc, char *argv[]);
+static int select_codec(int argc, char *argv[]);
+static int properties(int argc, char *argv[]);
+static int set_volume(int argc, char *argv[]);
+static int mute(int argc, char *argv[]);
+static int softvol(int argc, char *argv[]);
+static int open_pcm(int argc, char *argv[]);
+static int monitor(int argc, char *argv[]);
+
+
+static struct command {
+	const char *name;
+	int (*func)(int argc, char *arg[]);
+	const char *help;
+} commands[] = {
+	{ "get-codecs", get_codecs, "PCM_PATH", },
+	{ "list-pcms", list_pcms, "" },
+	{ "monitor", monitor, "" },
+	{ "mute", mute, "PCM_PATH y|n [y|n]" },
+	{ "open", open_pcm, "PCM_PATH" },
+	{ "properties", properties, "PCM_PATH" },
+	{ "select-codec", select_codec, "PCM_PATH CODEC" },
+	{ "set-volume", set_volume, "PCM_PATH N [N]" },
+	{ "softvol", softvol, "PCM_PATH y|n" },
+	{ 0 }
+};
+
+static const struct option long_options[] = {
```

Comment:
Place this structure in the main function. Because in case of more complicate subcommand it might happen that subcommand will have its own `getopt()`.

Reply:
see commit 4b97de6